profiler/profiler.go

```diff
@@ -23,13 +24,64 @@ var (
 
 	// A sink for emitted profile entries.
 	outputSink Sink
+
+	// Function call invokation overhead; calculated by calibrate() and triggered by init()
+	timeNowOverhead, timeSinceOverhead, deferredFnOverhead, fnCallOverhead time.Duration
 )
 
-// Time is an alias to time.Time(). It is exposed by this package so our injected
-// code does not need to import the time package which may cause conflict with
-// other imported/aliased packages.
-func Time() time.Time {
-	return time.Now()
+func init() {
+	calibrate()
+}
+
+// callibrate attempts to estimate the mean overhead for invoking time.Now(), time.Since(),
+// as well as the mean time spent in function guard code (stack setup, pushing/popping
+// registers, dealing with deferred calls e.t.c).
+//
+// Runtime overhead is generally in the nanosecond range but we need to
+// properly account for it when calculating the total time spent inside a
+// profiled function as it tends to skew our timing calculations when the profiled
+// function is invoked a large number of times.
+//
+// To calculate an estimate, we time N executions of each function and then just
+// calculate the mean execution time. As outliers may skew our results, using
+// the median execution time would be better but calculating it is not a trivial
+// operation due to the space and timing accurace required.
+func calibrate() {
+	// Benchmark function call time. We use a switch statement to ensure that
+	// the compiler will not inline this function (see https://github.com/golang/go/issues/12312)
+	fnCallBench := func(i int) {
+		switch i {
+		}
+	}
+	tick := time.Now()
+	for i := 0; i < numCalibrationCalls; i++ {
+		fnCallBench(i)
+	}
+	fnCallOverhead = time.Since(tick) / time.Duration(numCalibrationCalls)
```

Comment:
does golang do any JIT that would cause the first few invocations to skew slower? (I honestly don't know)